Running habushu with mvnd with many simultaneous executions results in errors ~10% of the time

When running a project with *+ simultaneous habushu executions, the following error is encountered:

```
[INFO] --- habushu:2.18.0:validate-pyenv-and-poetry (default-validate-pyenv-and-poetry) @ booz-allen-public-license-python ---
[INFO] Using Python 3.11.4 (managed by pyenv)
[INFO] Found Poetry 2.1.1
[INFO] Configuring Poetry to use the pyenv-activated Python binary...
[INFO] Executing Poetry command: poetry config --local virtualenvs.use-poetry-python false
[INFO] Checking for updates to poetry-monorepo-dependency-plugin...
[INFO] Executing Poetry command: poetry self show plugins
[ERROR] 
The Poetry configuration is invalid:
  - Either [project.name] or [tool.poetry.name] is required in package mode.
  - Either [project.version] or [tool.poetry.version] is required in package mode.
```

This is an improvement over 2.17.x, but still would be nice to make it work all the time.

## Steps to Recreate:
* Check out a project with a lot of habushu 2.18.0 projects and run `mvnd initialize`
   * you may need to turn caching off, if you are using it on the project: `-Dmaven.build.cache.skipCache=true`
* If you run a project with 5+ modules that all use habushu somewhat simultaneously, you'll hit this error about one in ten runs

## Test Steps
* Check out a project with a lot of habushu 2.18.0 projects and run `mvnd initialize`
   * you may need to turn caching off, if you are using it on the project: `-Dmaven.build.cache.skipCache=true`
* Update it to use habushu 2.18.1-SHAPSHOT (ensure you have built this locally as well)
* Run the command at least ten times and verify you don't see the error. 
* Run `poetry self remove poetry-monorepo-dependency-plugin`
* Run `mvnd initialize` again, verify no errors occur

--- a/habushu-maven-plugin/pom.xml
+++ b/habushu-maven-plugin/pom.xml
@@ -4,7 +4,7 @@
     <parent>
         <groupId>org.technologybrewery.habushu</groupId>
         <artifactId>habushu</artifactId>
-        <version>2.18.0</version>
+        <version>2.18.1-SNAPSHOT</version>
     </parent>
 
     <name>habushu::Maven Plugin for Python Projects</name>

--- a/habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/exec/PoetryCommandHelper.java
+++ b/habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/exec/PoetryCommandHelper.java
@@ -20,6 +20,7 @@ import org.apache.maven.artifact.versioning.DefaultArtifactVersion;
 import org.apache.maven.plugin.MojoExecutionException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.technologybrewery.habushu.HabushuException;
 
 /**
  * Facilitates the execution of Poetry commands.
@@ -36,6 +37,8 @@ public class PoetryCommandHelper {
 
     protected File workingDirectory;
 
+    protected volatile String showPluginsResult;
+    protected static final Object pluginSystemLock = new Object();
 
     public PoetryCommandHelper(File workingDirectory) {
         this.workingDirectory = workingDirectory;
@@ -236,26 +239,26 @@ public class PoetryCommandHelper {
         return result;
     }
 
-    private synchronized int performInstallPoetryPlugin(String name) throws MojoExecutionException {
+    private int performInstallPoetryPlugin(String name) throws MojoExecutionException {
         int result = EXIT_SUCCESS;
-        if (pluginNeedsInstalling(name)) {
-            List<String> args = new ArrayList<>();
-            args.add("self");
-            args.add("add");
-            args.add(name);
-            result = this.executeAndLogOutput(args);
-
+        List<String> args = new ArrayList<>();
+        args.add("self");
+        args.add("add");
+        args.add(name);
+        synchronized (pluginSystemLock) {
+            if (pluginNeedsInstalling(name)) {
+                result = this.executeAndLogOutput(args);
+            }
         }
+
+        // un-cache known set of plugins:
+        showPluginsResult = null;
 
         return result;
     }
 
     private boolean pluginNeedsInstalling(String name) throws MojoExecutionException {
-        List<String> args = new ArrayList<>();
-        args.add("self");
-        args.add("show");
-        args.add("plugins");
-        String showPluginsResult =  this.execute(args);
+        executeShowPlugins();
 
         String pluginNameWithoutVersion = name;
         if (name.contains(VERSION_DELIMITER)) {
@@ -263,6 +266,30 @@ public class PoetryCommandHelper {
         }
 
         return !showPluginsResult.contains(pluginNameWithoutVersion);
+    }
+
+    private void executeShowPlugins() throws MojoExecutionException {
+        if (StringUtils.isBlank(showPluginsResult)) {
+            List<String> args = new ArrayList<>();
+            args.add("self");
+            args.add("show");
+            args.add("plugins");
+            try {
+                    if (StringUtils.isBlank(showPluginsResult)) {
+                        showPluginsResult = this.execute(args);
+                    }
+            } catch (HabushuException e) {
+                logger.info("Plugin status could not be determined. This is normally a race condition on another"
+                        + " thread touching poetry's underlying pyproject.toml, trying one more time...");
+                // let's be more careful round two and assume that
+                synchronized (pluginSystemLock) {
+                    if (StringUtils.isBlank(showPluginsResult)) {
+                        showPluginsResult = this.execute(args);
+                    }
+                }
+
+            }
+        }
     }
 
     protected ProcessExecutor createPoetryExecutor(List<String> arguments) {

--- a/habushu-mixology-consumer/pom.xml
+++ b/habushu-mixology-consumer/pom.xml
@@ -4,7 +4,7 @@
     <parent>
         <groupId>org.technologybrewery.habushu</groupId>
         <artifactId>habushu</artifactId>
-        <version>2.18.0</version>
+        <version>2.18.1-SNAPSHOT</version>
     </parent>
 
     <name>habushu::Mixing Drinks with Habushu Consumer</name>

--- a/habushu-mixology-consumer/pyproject.toml
+++ b/habushu-mixology-consumer/pyproject.toml
@@ -2,7 +2,7 @@
 
 [project]
 name = "habushu_mixology_consumer"
-version = "2.18.0"
+version = "2.18.1.dev"
 description = "Example of how to use the habushu-maven-plugin to consume other Habushu modules"
 authors = [{name = "Eric Konieczny", email = "ekoniec1@gmail.com"}]
 license = "MIT License"

--- a/habushu-mixology/pom.xml
+++ b/habushu-mixology/pom.xml
@@ -5,7 +5,7 @@
     <parent>
         <groupId>org.technologybrewery.habushu</groupId>
         <artifactId>habushu</artifactId>
-        <version>2.18.0</version>
+        <version>2.18.1-SNAPSHOT</version>
     </parent>
 
     <name>habushu::Mixing Drinks with Habushu</name>

--- a/habushu-mixology/pyproject.toml
+++ b/habushu-mixology/pyproject.toml
@@ -2,7 +2,7 @@
 
 [project]
 name = "habushu_mixology"
-version = "2.18.0"
+version = "2.18.1.dev"
 description = "Example of how to use the habushu-maven-plugin"
 authors = [{name = "Eric Konieczny", email = "ekoniec1@gmail.com"}]
 license = "MIT License"

--- a/pom.xml
+++ b/pom.xml
@@ -9,7 +9,7 @@
     </parent>
 
     <groupId>org.technologybrewery.habushu</groupId>
-    <version>2.18.0</version>
+    <version>2.18.1-SNAPSHOT</version>
     <artifactId>habushu</artifactId>
     <packaging>pom</packaging>
 

